divi/src/BlockMemoryPoolTransactionCollector.cpp

```diff
@@ -18,10 +20,6 @@ extern CCoinsViewCache* pcoinsTip;
 extern bool fDebug;
 
 class CScriptCheck;
-unsigned int GetLegacySigOpCount(const CTransaction& tx);
-unsigned int GetP2SHSigOpCount(const CTransaction& tx, const CCoinsViewCache& mapInputs);
-bool CheckInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& view, bool fScriptChecks, unsigned int flags, bool cacheStore, std::vector<CScriptCheck>* pvChecks = NULL );
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo& txundo, int nHeight);
```

Comment:
Instead of including main to get access to these functions, please separate them out into a separate file to avoid recompiling main each time.

Reply:
I personally disagree with this, as I'd rather have longer compilation time than duplicated forward declarations (which may introduce bugs, although the risk is admittedly rather low).  But as this is your decision, I've reverted that part of the change now and updated the forward declarations instead of including `main.h`.

It may make sense to move some of these functions to a separate file and header as you say (e.g. upstream Bitcoin has long created a `validation` file), but that would be a separate PR.  I can do that before this if you want, or we just keep it in mind for after.

Reply:
I agree with your concern over duplicated forward declarations of functions. I would also rather include the header directly __if there were no alternative allowing moving the definition out of main__. Moving the necessary data into header/source pairs ought to be preferred. There's no need to pay for a long compilation time where a shorter one with the same amount of bugs would do. If you have concerns over potential name resolution I'm all for placing them either in their own class or namespace to avoid those potential sources of conflict.

Reply:
Makes sense.  I saw you did some code moves recently.  I will rebase this PR on top of them, and then see what to do here.  In any case, I think those moves can/should be separate PRs merged before this one then.

Reply:
Rebased & thanks to your work, this is now obsolete :+1: